File: test/domain/scoring/scoring_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:word_search_master/domain/scoring/score_event.dart';
import 'package:word_search_master/domain/scoring/scoring.dart';

void main() {
  group('base points', () {
    test('a word is worth 10 points per grapheme at combo 1', () {
      expect(Scoring.wordScore(graphemeCount: 5, consecutiveCorrect: 1), 50);
      expect(Scoring.wordScore(graphemeCount: 2, consecutiveCorrect: 1), 20);
    });

    test('graphemes are the unit, so a Hindi word scores by aksharas', () {
      // "पानी" is 2 graphemes and 4 code points. Scoring the code points would
      // pay double for every matra.
      expect(Scoring.wordScore(graphemeCount: 2, consecutiveCorrect: 1), 20);
    });

    test('a degenerate grapheme count scores nothing rather than negative', () {
      expect(Scoring.wordScore(graphemeCount: 0, consecutiveCorrect: 3), 0);
      expect(Scoring.wordScore(graphemeCount: -4, consecutiveCorrect: 3), 0);
    });
  });

  group('the combo ladder', () {
    test('runs x1.0, x1.2, x1.4, x1.6, x1.8, x2.0', () {
      const expected = [1.0, 1.2, 1.4, 1.6, 1.8, 2.0];
      for (var step = 1; step <= 6; step++) {
        expect(
          Scoring.comboMultiplier(step),
          closeTo(expected[step - 1], 1e-9),
          reason: 'combo $step',
        );
      }
    });

    test('caps at the sixth consecutive word', () {
      for (final step in [6, 7, 12, 500]) {
        expect(Scoring.comboMultiplier(step), 2.0, reason: 'combo $step');
        expect(
          Scoring.wordScore(graphemeCount: 5, consecutiveCorrect: step),
          100,
        );
      }
    });

    test('the integer ladder is exactly 10x the displayed multiplier', () {
      // Scoring multiplies by these integers rather than by the doubles, so
      // Dart and the TypeScript port cannot drift apart on float rounding —
      // and one point of drift means a rejected submission.
      for (var step = 1; step <= 6; step++) {
        expect(
          Scoring.comboPointsPerGrapheme[step - 1],
          (Scoring.comboMultiplier(step) * Scoring.basePointsPerGrapheme)
              .round(),
        );
      }
      expect(Scoring.comboPointsPerGrapheme, [10, 12, 14, 16, 18, 20]);
    });

    test('the full ladder scores as specified, word by word', () {
      const word = 5;
      const expectedPerWord = [50, 60, 70, 80, 90, 100, 100];

      var running = 0;
      for (var i = 0; i < expectedPerWord.length; i++) {
        running += expectedPerWord[i];
        final events = [
          for (var w = 0; w <= i; w++) const WordFound(graphemeCount: word),
        ];
        expect(
          Scoring.computeScore(events),
          running,
          reason: '${i + 1} consecutive words',
        );
      }
    });

    test('a wrong selection resets the combo to zero', () {
      // Five in a row, then a miss, then one more: the sixth word scores at
      // x1.0 again, not x2.0.
      final events = <ScoreEvent>[
        for (var i = 0; i < 5; i++) const WordFound(graphemeCount: 5),
        const WrongSelection(),
        const WordFound(graphemeCount: 5),
      ];

      // 50 + 60 + 70 + 80 + 90 = 350, then 50 at the reset combo.
      expect(Scoring.computeScore(events), 400);
    });

    test('a wrong selection costs no points itself', () {
      const found = [WordFound(graphemeCount: 4)];
      const withMiss = [WordFound(graphemeCount: 4), WrongSelection()];

      expect(Scoring.computeScore(withMiss), Scoring.computeScore(found));
    });

    test('a hint does NOT reset the combo', () {
      final withHint = <ScoreEvent>[
        const WordFound(graphemeCount: 5),
        const HintUsed(),
        const WordFound(graphemeCount: 5),
      ];
      // Second word still scores at combo 2 (60), minus the 25 penalty.
      expect(Scoring.computeScore(withHint), 50 + 60 - 25);
    });

    test('maxComboIn reports the longest run', () {
      final events = <ScoreEvent>[
        const WordFound(graphemeCount: 3),
        const WordFound(graphemeCount: 3),
        const WrongSelection(),
        const WordFound(graphemeCount: 3),
        const WordFound(graphemeCount: 3),
        const WordFound(graphemeCount: 3),
      ];

      expect(Scoring.maxComboIn(events), 3);
    });
  });

  group('hints', () {
    test('each hint costs a flat 25 points', () {
      final one = <ScoreEvent>[
        const WordFound(graphemeCount: 10),
        const HintUsed(),
      ];
      final two = <ScoreEvent>[
        const WordFound(graphemeCount: 10),
        const HintUsed(),
        const HintUsed(),
      ];

      expect(Scoring.computeScore(one), 100 - 25);
      expect(Scoring.computeScore(two), 100 - 50);
    });

    test('the penalty is not scaled by the combo', () {
      final early = <ScoreEvent>[
        const HintUsed(),
        for (var i = 0; i < 6; i++) const WordFound(graphemeCount: 5),
      ];
      final late = <ScoreEvent>[
        for (var i = 0; i < 6; i++) const WordFound(graphemeCount: 5),
        const HintUsed(),
      ];

      expect(Scoring.computeScore(early), Scoring.computeScore(late));
    });

    test('the score floors at zero — hints never push a player negative', () {
      final events = <ScoreEvent>[
        const WordFound(graphemeCount: 2),
        for (var i = 0; i < 10; i++) const HintUsed(),
      ];

      expect(Scoring.computeScore(events), 0);
    });

    test('hintsIn counts only paid hints', () {
      final events = <ScoreEvent>[
        const HintUsed(),
        const WordFound(graphemeCount: 3),
        const HintUsed(),
        const WrongSelection(),
      ];

      expect(Scoring.hintsIn(events), 2);
    });
  });

  group('stars — relaxed mode', () {
    test('3 for no hints, 2 for one, 1 for two or more', () {
      expect(Scoring.computeStars(hintsUsed: 0), 3);
      expect(Scoring.computeStars(hintsUsed: 1), 2);
      expect(Scoring.computeStars(hintsUsed: 2), 1);
      expect(Scoring.computeStars(hintsUsed: 9), 1);
    });

    test('never returns 0 — a completed level always earns something', () {
      for (var hints = 0; hints < 50; hints++) {
        expect(Scoring.computeStars(hintsUsed: hints), inInclusiveRange(1, 3));
      }
    });

    test('a negative hint count is treated as none', () {
      expect(Scoring.computeStars(hintsUsed: -3), 3);
    });

    test('stars cannot depend on time — there is no time to depend on', () {
      // Enforced by the signature rather than by convention: relaxed mode is
      // the product default and timing pressure is what drives this game's
      // audience away (Ch01/Ch02). Blitz (v1.2) gets its own function.
      //
      // Two levels with identical hints but wildly different amounts of play
      // are indistinguishable to this function.
      expect(
        Scoring.computeStars(hintsUsed: 1),
        Scoring.computeStars(hintsUsed: 1),
      );
    });
  });

  group('purity', () {
    test('computeScore is deterministic over repeated calls', () {
      final events = <ScoreEvent>[
        const WordFound(graphemeCount: 5),
        const WordFound(graphemeCount: 4),
        const WrongSelection(),
        const WordFound(graphemeCount: 3),
        const HintUsed(),
      ];

      final first = Scoring.computeScore(events);
      for (var i = 0; i < 100; i++) {
        expect(Scoring.computeScore(events), first);
      }
    });

    test('computeScore does not mutate its input', () {
      final events = <ScoreEvent>[
        const WordFound(graphemeCount: 5),
        const WrongSelection(),
      ];
      final snapshot = List<ScoreEvent>.from(events);

      Scoring.computeScore(events);

      expect(events, snapshot);
    });

    test('an empty level scores zero', () {
      expect(Scoring.computeScore(const []), 0);
      expect(Scoring.maxComboIn(const []), 0);
      expect(Scoring.hintsIn(const []), 0);
    });
  });

  group('the spec header worked example', () {
    test('scores 103 with 2 stars, exactly as documented', () {
      // This case is quoted verbatim in the scoring.dart spec header and is
      // the cross-language parity fixture for the TypeScript port in P14. If
      // this number changes, the port and the spec must change with it.
      final events = <ScoreEvent>[
        const WordFound(graphemeCount: 5),
        const WordFound(graphemeCount: 4),
        const WrongSelection(),
        const WordFound(graphemeCount: 3),
        const HintUsed(),
      ];

      expect(Scoring.computeScore(events), 103);
      expect(Scoring.computeStars(hintsUsed: Scoring.hintsIn(events)), 2);
      expect(Scoring.maxComboIn(events), 2);
    });

    test('the spec version is pinned so a stale client is detectable', () {
      expect(Scoring.specVersion, 1);
    });
  });
}

File: functions/test/scoring.test.ts
import { describe, expect, it } from 'vitest';

import {
  COMBO_POINTS_PER_GRAPHEME,
  HINT_PENALTY,
  MAX_COMBO_STEP,
  SPEC_VERSION,
  comboMultiplier,
  computeScore,
  computeStars,
  hintsIn,
  maxComboIn,
  wordScore,
  wordsFoundIn,
  type ScoreEvent,
} from '../src/scoring';

const found = (g: number): ScoreEvent => ({ t: 'w', g });
const wrong: ScoreEvent = { t: 'x' };
const hint: ScoreEvent = { t: 'h' };

describe('the spec header worked example', () => {
  // Verbatim from `lib/domain/scoring/scoring.dart`'s header. This is the
  // cross-language parity fixture the spec names, so it is asserted here as
  // well as in the Dart tests and in the 200-case fixture.
  const events = [found(5), found(4), wrong, found(3), hint];

  it('scores 103', () => {
    expect(computeScore(events)).toBe(103);
  });

  it('awards 2 stars for the one hint', () => {
    expect(computeStars(hintsIn(events))).toBe(2);
  });

  it('walks the ladder 50 -> 98 -> 98 -> 128 -> 103', () => {
    expect(computeScore(events.slice(0, 1))).toBe(50);
    expect(computeScore(events.slice(0, 2))).toBe(98);
    expect(computeScore(events.slice(0, 3))).toBe(98);
    expect(computeScore(events.slice(0, 4))).toBe(128);
    expect(computeScore(events.slice(0, 5))).toBe(103);
  });
});

describe('the combo ladder', () => {
  it('is the integer table, not a float multiply', () => {
    expect(COMBO_POINTS_PER_GRAPHEME).toEqual([10, 12, 14, 16, 18, 20]);
    // Every entry must be an exact integer: the whole reason the table exists
    // is that `10 * 1.2` is not reliably 12 in two languages.
    for (const points of COMBO_POINTS_PER_GRAPHEME) {
      expect(Number.isInteger(points)).toBe(true);
    }
  });

  it('caps at step 6 and never grows past it', () => {
    expect(wordScore(1, MAX_COMBO_STEP)).toBe(20);
    expect(wordScore(1, MAX_COMBO_STEP + 1)).toBe(20);
    expect(wordScore(1, 1000)).toBe(20);
  });

  it('treats a combo below 1 as step 1', () => {
    expect(wordScore(3, 0)).toBe(30);
    expect(wordScore(3, -5)).toBe(30);
  });

  it('exposes display multipliers that match the table exactly', () => {
    expect([1, 2, 3, 4, 5, 6].map(comboMultiplier)).toEqual([
      1.0, 1.2, 1.4, 1.6, 1.8, 2.0,
    ]);
  });
});

describe('rules that a naive port would get wrong', () => {
  it('does not reset the combo on a hint', () => {
    // 3*10 + 3*12 = 66, minus one hint.
    expect(computeScore([found(3), hint, found(3)])).toBe(66 - HINT_PENALTY);
  });

  it('resets the combo on a wrong selection', () => {
    expect(computeScore([found(3), wrong, found(3)])).toBe(60);
  });

  it('clamps the score at zero rather than going negative', () => {
    expect(computeScore([hint, hint, hint])).toBe(0);
    expect(computeScore([found(2), hint, hint])).toBe(0);
  });

  it('scores a non-positive grapheme count as zero but still advances the combo', () => {
    // 0 then -3 pay nothing, but the third word is the THIRD in the streak.
    expect(computeScore([found(0), found(-3), found(2)])).toBe(2 * 14);
  });
});

describe('stars', () => {
  it('follows the relaxed-mode table', () => {
    expect(computeStars(0)).toBe(3);
    expect(computeStars(1)).toBe(2);
    expect(computeStars(2)).toBe(1);
    expect(computeStars(9)).toBe(1);
  });

  it('treats a negative hint count as zero hints', () => {
    expect(computeStars(-1)).toBe(3);
  });
});

describe('event summaries', () => {
  it('counts hints, words and the longest streak', () => {
    const events = [found(2), found(2), wrong, found(2), hint, found(2)];
    expect(hintsIn(events)).toBe(1);
    expect(wordsFoundIn(events)).toBe(4);
    expect(maxComboIn(events)).toBe(2);
  });

  it('reports a zero max combo for a replay with no words', () => {
    expect(maxComboIn([wrong, hint, wrong])).toBe(0);
  });

  it('does not break a streak across a hint', () => {
    expect(maxComboIn([found(2), hint, found(2), hint, found(2)])).toBe(3);
  });
});

it('declares the same spec version as the Dart side', () => {
  expect(SPEC_VERSION).toBe(1);
});
